torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+    To help the model "see" the whole image, we use positional embeddings. If your image
+    was tile-cropped, then you need to use tile positional embeddings:
+
+    - token_pos_embedding (tiled): ``torchtune.models.clip._position_embeddings.TiledTokenPositionalEmbedding``
+    - pre_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+    - post_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+
+    Otherwise, pre and post tile_pos_embed should be None. For the tokens, all you need is a simple
+    token positional embedding:
+
+    - token_pos_embedding: ``torchtune.models.clip._position_embeddings.TokenPositionalEmbedding``
+
+    All images will be considered as a stack of tiles, even if your image was not tile-cropped. In such cases,
+    your image would be composed of a single tile.
+
+    Args:
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+        num_layers (int): The number of transformer layers.
+        layer (nn.Module): The transformer layer module.
+        token_pos_embedding (nn.Module): The token positional embedding module.
+        pre_tile_pos_embed (Optional[nn.Module]): The pre-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        post_tile_pos_embed (Optional[nn.Module]): The post-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        cls_projection (Optional[nn.Module]): The CLS projection module. It should take an input tensor
+            of shape (bsz * n_tiles, n_tokens, embed_dim) and output a tensor of shape
+            (bsz * n_tiles, cls_output_dim). If provided, only the CLS token projection will be
+            outputted, instead of all tokens.
+        indices_return_hidden (Optional[List[int]]): The indices of hidden layers to return.
+            If provided, it will return the intermediate results of the transformer layers
+            before they go through a next layer. For example, indices_return_hidden = [0, 3] will
+            return the tokens before they go through the first and fourth layers.
+        patch_size (int): The size of each patch. Used to divide the tiles into patches.
+            E.g. for patch_size = 40, a tile of shape (400, 400) will have 10x10 grid of patches
+            with shape (40, 40) each.
+        embed_dim (int): The dimensionality of each patch embedding (token).
+        in_channels (int): The number of image input channels.
+    """
+
+    def __init__(
+        self,
+        patch_grid_size: int,
+        num_layers: int,
+        layer: nn.Module,
+        token_pos_embedding: nn.Module,
+        pre_tile_pos_embed: Optional[nn.Module] = None,
+        post_tile_pos_embed: Optional[nn.Module] = None,
+        cls_projection: Optional[nn.Module] = None,
```

Comment:
Another thing you could consider is just using nn.Identity in the builder as needed rather than having all these be `Optional` and inside of if statements in the model forward

Reply:
this is neat, i just dont know if as an user, this would confuse me. 

what is more intuitive, in your opinion?

Option 1: Have if statement, but its clear that this is not executed if not defined
```
if pre_tile_pos_embed:
    x = pre_tile_pos_embed(x)
```
    
 or
 
 Option 2: Its always executed, but user has to know that pre_tile_pos_embed is just an identify operation that doesnt do anything
`x = pre_tile_pos_embed(x)`

Personally, I think that option 1 is easier to understand, even though it is a bit more bloated

Reply:
> Its always executed, but user has to know that pre_tile_pos_embed is just an identify operation that doesnt do anything

You can always take an nn.Module here and have the builder send in an nn.Identity if there's no embedding module (though better if we leave that as the default). But in either ways I think you need to trace through the builder to know how the component is setup and in one case we avoid an additional if branch in the code

Reply:
Yeah I would go with option 2 personally. Mainly we want (a) forward to be as clean as possible and (b) the class to be really generic with the potential for multiple different instantiations via builders. We're already typing the embeddingsas nn.Module anyways so I don't think we are introducing additional ambiguity by doing this.

Reply:
Discussed a bit offline -- I goofed here and thought that `nn.Identity` can take in unused arguments in forward, which it cannot. So this won't work for mocking `TilePositionalEmbedding`, which also takes in `aspect_ratio` in its forward. As @felipemello1 suggested, we could get around this by defining some wrapper to take in arbitrary arguments and return only the first one. But in that case we are just introducing extra indirection and it may be better to just leave this as is.